datalad/plugin/addurls.py

```diff
@@ -0,0 +1,654 @@
+        super(RepFormatter, self).__init__(*args, **kwargs)
+        self.repeats = {}
+        self.repindex = 0
+
+    def format(self, *args, **kwargs):
+        self.repindex = 0
+        result = super(RepFormatter, self).format(*args, **kwargs)
+        if result in self.repeats:
+            self.repindex = self.repeats[result] + 1
+            self.repeats[result] = self.repindex
+            result = super(RepFormatter, self).format(*args, **kwargs)
+        else:
+            self.repeats[result] = 0
+        return result
+
+    def get_value(self, key, args, kwargs):
+        args[0]["_repindex"] = self.repindex
+        return super(RepFormatter, self).get_value(key, args, kwargs)
+
+
+def clean_meta_args(args):
+    """Prepare formatted metadata arguments to be passed to git-annex.
+
+    Parameters
+    ----------
+    args : iterable of str
+        Formatted metadata arguments for 'git-annex metadata --set'.
+
+    Returns
+    -------
+    Generator that yields processed arguments (str).
+    """
+    for arg in args:
+        parts = [x.strip() for x in arg.split("=", 1)]
+        if len(parts) == 2:
+            if not parts[0]:
+                raise ValueError("Empty field name")
+            field, value = parts
+        else:
+            raise ValueError("meta argument isn't in 'field=value' format")
+
+        if not value:
+            # The `url_file` may have an empty value.
+            continue
+
+        yield field + "=" + value
+
+
+def get_subpaths(filename):
+    """Convert "//" marker in `filename` to a list of subpaths.
+
+    >>> get_subpaths("p1/p2//p3/p4//file")
+    ('p1/p2/p3/p4/file', ['p1/p2', 'p1/p2/p3/p4'])
+
+    Note: With Python 3, the subpaths could be generated with
+
+        itertools.accumulate(filename.split("//")[:-1], os.path.join)
+
+    Parameters
+    ----------
+    filename : str
+        File name with "//" marking subpaths.
+
+    Returns
+    -------
+    A tuple of the filename with any "//" collapsed to a single
+    separator and a list of subpaths (str).
+    """
+    if "//" not in filename:
+        return filename, []
+
+    spaths = []
+    for part in filename.split("//")[:-1]:
+        path = os.path.join(*(spaths + [part]))
+        spaths.append(path)
+    return filename.replace("//", os.path.sep), spaths
+
+
+def is_legal_metafield(name):
+    """Test whether `name` is a valid metadata field.
+
+    The set of permitted characters is taken from git-annex's
+    MetaData.hs:legalField.
+    """
+    legal = re.compile(r"[a-zA-Z0-9][a-zA-Z0-9_.-]*\Z")
+    return True if legal.match(name) else False
+
+
+def filter_legal_metafield(fields):
+    """Remove illegal names from `fields`.
+    """
+    legal = []
+    for field in fields:
+        if is_legal_metafield(field):
+            legal.append(field)
+        else:
+            lgr.debug("%s is not a valid metadata field name; dropping",
+                      field)
+    return legal
```

Comment:
I guess the only purpose of having it this verbose is debugging, since otherwise it is pretty much the name of the function so could have been used in the code as is without a dedicated helper:
```python
return filter(is_legal_metafield, fields)
```
;-)  (yeap, would be a generator in python3, only for good) so might be worth adding a note for such creatures as I am that this is for extended debugging?

Reply:
> I guess the only purpose of having it this verbose is debugging

Exactly.

> so might be worth adding a note for such creatures as I am that this is for extended debugging?

In my eyes, the presence of the debugging statement pretty much serves as the note, but sure, I'll add a comment.